twilight-gateway/src/shard.rs

```diff
@@ -669,455 +514,95 @@ impl<Q> Shard<Q> {
-    /// [`next_message`]: Self::next_message
-    pub async fn next_event(&mut self) -> Result<Event, ReceiveMessageError> {
-        loop {
-            match self.next_message().await? {
-                Message::Close(frame) => return Ok(Event::GatewayClose(frame)),
-                Message::Text(text) => match crate::parse(text, self.config.event_types()) {
-                    Ok(Some(event)) => return Ok(event.into()),
-                    #[allow(clippy::redundant_closure_for_method_calls)]
-                    Err(e)
-                        if e.source()
-                            .and_then(|source| source.downcast_ref::<UnknownEventError>())
-                            .is_none() =>
-                    {
-                        return Err(e)
-                    }
-                    _ => {}
-                },
-            }
-        }
-    }
-
-    /// Wait for the next raw message from the websocket connection.
-    ///
-    /// # Errors
-    ///
-    /// Returns a [`ReceiveMessageErrorType::Compression`] error type if the
-    /// message payload failed to decompress.
-    ///
-    /// Returns a [`ReceiveMessageErrorType::FatallyClosed`] error type if the
-    /// shard was closed due to a fatal error, such as invalid authorization.
-    ///
-    /// Returns a [`ReceiveMessageErrorType::Process`] error type if the shard
-    /// failed to internally process a received event.
-    ///
-    /// Returns a [`ReceiveMessageErrorType::Reconnect`] error type if the shard
-    /// failed to reconnect to the gateway. This isn't a fatal error and can be
-    /// retried.
-    ///
-    /// Returns a [`ReceiveMessageErrorType::SendingMessage`] error type if the
-    /// shard failed to send a message to the gateway, such as a heartbeat.
-    #[tracing::instrument(fields(id = %self.id), name = "shard", skip(self))]
-    pub async fn next_message(&mut self) -> Result<Message, ReceiveMessageError> {
-        /// Actions the shard might take.
-        enum Action {
-            /// Close the gateway connection with this close frame.
-            Close(CloseFrame<'static>),
-            /// Send this command to the gateway.
-            Command(String),
-            /// Send a heartbeat command to the gateway.
-            Heartbeat,
-            /// Identify with the gateway.
-            Identify,
-            /// Handle this incoming gateway message.
-            Message(Option<Result<WebsocketMessage, WebsocketError>>),
-        }
 
-        match self.status {
-            ConnectionStatus::Disconnected {
-                close_code,
-                reconnect_attempts,
-            } => {
-                // The shard should should only reconnect after the gateway
-                // closes the underlying TCP connection.
-                if self.connection.is_none() {
-                    self.reconnect(close_code, reconnect_attempts).await?;
+    /// Send and flush the pending message.
+    fn poll_flush_pending(
```

Comment:
Not a fan of the massive amounts of unwrapping here, it just seems a bit ugly. Perhaps unwrap it once and reuse that?

Reply:
Borrow checker doesn't like that.

Reply:
Ah yeah I've seen that before. The "fix" was to assign all fields you're going to use to local variables at the beginning, which I guess isn't prettier...